### PR TITLE
Move async logic out of the client, address #248

### LINKED ISSUE #248: Implement asynchronous provision and deprovision
According to the consensus established for #49

### PR DESCRIPTION
**Pull request by vaikas-google**
Fixes #248 
Follow on work in consequent PRs are: 
 - Move to a single queue solution
 - Refactor the delete code to a method like pollInstance
 - write complete unit tests for corner cases

### CODE REVIEW

File: pkg/brokerapi/openservicebroker/exponential_backoff_poller.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+			if !cb(instance, resp) {
+				glog.V(3).Infof("Stopping polling for instance %q", instanceID)
+				return
+			}
+		}
+		time.Sleep(p.d)
+		p.d = 2 * p.d
+		if p.d > max {
+			p.d = max
+		}
```
Comment (jmrodri) [inline pkg/brokerapi/openservicebroker/exponential_backoff_poller.go]: EXCELLENT!

File: pkg/brokerapi/openservicebroker/exponential_backoff_poller.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1"
+	"github.com/kubernetes-incubator/service-catalog/pkg/brokerapi"
+)
+
+const (
+	// Max specifies the maximum duration we're going to back off up to
+	max = 1 * time.Hour
```
Comment (jmrodri) [inline pkg/brokerapi/openservicebroker/exponential_backoff_poller.go]: 1 hour is acceptable for our broker. We have some provision calls that could take 10-30 minutes to complete. Some of the more complex ones would require more time. Hopefully we'll never exceed an hour :)

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go [hunk 3]
```diff
@@ -40,9 +40,10 @@ const (
 	serviceInstanceAsyncFormatString       = "%s/v2/service_instances/%s?accepts_incomplete=true"
 	serviceInstanceDeleteFormatString      = "%s/v2/service_instances/%s?service_id=%s&plan_id=%s"
 	serviceInstanceDeleteAsyncFormatString = "%s/v2/service_instances/%s?service_id=%s&plan_id=%s&accepts_incomplete=true"
-	pollingFormatString                    = "%s/v2/service_instances/%s/last_operation"
+	pollingFormatString                    = "%s/v2/service_instances/%s/last_operation?%s"
```
Comment (jmrodri) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: :+1: including the operation as part of the string. What happens if the provision response does NOT return an operation?  Will the dangling '?' be a problem?

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go [hunk 4]
```diff
@@ -299,6 +299,76 @@ func (c *openServiceBrokerClient) DeleteServiceBinding(instanceID, bindingID, se
...
+// According to the spec, ServiceID and PlanID should be included, so fail requests
+// without them as it indicates programming error on our part.
+func createPollParameters(req *brokerapi.LastOperationRequest) (string, error) {
+	if req.ServiceID == "" {
+		return "", fmt.Errorf("LastOperationRequest is missing service_id")
+	}
+	if req.PlanID == "" {
+		return "", fmt.Errorf("LastOperationRequest is missing plan_id")
+	}
+
```
Comment (jmrodri) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: I thought service_id and plan_id were optional.

> Note: Although the request query parameters service_id and plan_id are not required, the platform should include them on all last_operation requests it makes to service brokers.

The above code would return an error because they are empty, and would not add the operation.

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go (hunk 3 above)
Comment (vaikas-google) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: No, because as per spec, the platform will send plan_id and service_id. So, really only the operator is optional as per spec.

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go (hunk 4 above)
Comment (vaikas-google) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: Yes, but because we control the platform we'll always send them. They are always present, so from our code they will always be present and missing them is an error. I figured that missing them signifies coding error on our part. Thoughts?

File: pkg/brokerapi/openservicebroker/exponential_backoff_poller.go (hunk 1 above)
Comment (vaikas-google) [inline pkg/brokerapi/openservicebroker/exponential_backoff_poller.go]: :+1:

File: pkg/brokerapi/openservicebroker/exponential_backoff_poller.go (hunk 2 above)
Comment (vaikas-google) [inline pkg/brokerapi/openservicebroker/exponential_backoff_poller.go]: :) I have this hazy recollection that some brokers can take hours. But, we can tweak this as necessary... Thanks for the datapoint!

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go (hunk 3 above)
Comment (vaikas-google) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: What I meant to say is: as per spec, they _should_ be sent and since we control the code for sending them, they will be always sent in our controller.

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -665,18 +675,49 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) {
...
+			// Start polling this Service Instance
+			lastOperationRequest := &brokerapi.LastOperationRequest{
+				ServiceID: serviceClass.OSBGUID,
+				PlanID:    servicePlan.OSBGUID,
+			}
+			if response.Operation != "" {
+				lastOperationRequest.Operation = *instance.Spec.OSBLastOperation
+			}
+
+			err = c.poller.CreatePoller(brokerClient, instance, lastOperationRequest, c.handleLastOperationResponse)
```
Comment (pmorie) [inline pkg/controller/controller.go]: Chainsaw level: I would rather re-queue the instance to be re-reconciled and have logic that polled the endpoint here.  If we run a goroutine per async provision we could potentially exhaust our goroutines.  So, it's better to use a requeue mechanism than to poll in a goroutine.  @sjenning's PR #648 adds the use of a work queue that we could use for instances (and later bindings if/when we support async binding).

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -739,6 +780,65 @@ func findServicePlan(name string, plans []v1alpha1.ServicePlan) *v1alpha1.Servic
...
+		glog.Warningf("Got invalid state in LastOperationResponse: %q", resp.State)
+	}
+	return false
+}
+
+// setAsyncOperationOngoing updates the
+// InstanceConditionAsyncOperationInProgress to the specified status.
+func (c *controller) setAsyncOperationOngoing(instance *v1alpha1.Instance, status bool) {
+	if status {
+		c.updateInstanceCondition(
```
Comment (pmorie) [inline pkg/controller/controller.go]: I think we should just set the ready condition to `False` while an async provision is going, and use a descriptive reason/message instead of an additional condition.  Thoughts, @jwforres ?

Comment (jwforres) [inline pkg/controller/controller.go]: That is what we were expecting, if it is still provisioning, then its not Ready.  It would be really strange to me as an API consumer to have it tell me something was Ready when its not...

File: pkg/brokerapi/openservicebroker/exponential_backoff_poller.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+// ExponentialBackoffPoller implements a (./pkg/brokerapi).Poller with exponential backoff.
+// TODO(vaikas): Add cancel, or use queues.
```
Comment (arschles) [inline pkg/brokerapi/openservicebroker/exponential_backoff_poller.go]: a `context` would be a simple, easy way to add cancellation. I'd prefer that over the more complex queue mechanism fwiw

**Comment by arschles** [comment]
@vaikas-google generally, :+1: on moving the polling logic out of the broker client and into the controller. looks like this is going in the right direction

File: pkg/brokerapi/openservicebroker/exponential_backoff_poller.go (hunk 7 above)
Comment (pmorie) [inline pkg/brokerapi/openservicebroker/exponential_backoff_poller.go]: @arschles we're going to get the queue mechanism anyway in #648

File: pkg/brokerapi/fake/fake_poller.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+type Poller struct {
+}
+
+// NewPoller creates a poller that responds as specified
+func NewPoller() brokerapi.Poller {
+	return &Poller{}
+}
+
+// CreatePoller starts polling the specified last operation endpoint.
+func (p *Poller) CreatePoller(brokerapi.BrokerClient, *v1alpha1.Instance, *brokerapi.LastOperationRequest, brokerapi.PollerCallback) error {
```
Comment (vaikas-google) [inline pkg/brokerapi/fake/fake_poller.go]: Probably add to a queue, point is that in order for a 'follower' to update the condition, these are necessary to be kept for a context.

File: pkg/controller/controller.go (hunk 5 above)
Comment (vaikas-google) [inline pkg/controller/controller.go]: Yes, hence the high level question about using queues. Now if that goes in, I can use that.

File: pkg/controller/controller.go (hunk 6 above)
Comment (vaikas-google) [inline pkg/controller/controller.go]: I think trying to use a reason as a string to indicate state is a bit janky, but if we try to shove everything into a single condition, I'm fine with that. What is the reason for not adding another condition? By encoding the provisioning state into a condition seems janky. It seems like we'll end up in a state where the condition of being 'false' also requires parsing the string for some convention to figure out that there's an ongoing async operation.

Comment (vaikas-google) [inline pkg/controller/controller.go]: @jwforres are you objecting to adding another condition to indicate that there's an ongoing asynchronous operation, or just that we'll flip this False? and once the async operation finishes we'll set it to true? Or something else?

File: pkg/controller/controller.go (hunk 5 above)
Comment (vaikas-google) [inline pkg/controller/controller.go]: ** Not arguing against moving to queues** Just curious...
Can you say more about exhausting goroutines? I'm genuinely curious about what the limits here are so if you have pointers to what's sane and what's not, I'd love to see. My understanding was that goroutines are extremely light weight.

**Comment by pmorie** [comment]
@vaikas-google you should be able to rebase onto #648 now

File: pkg/controller/controller.go (hunk 6 above)
Comment (arschles) [inline pkg/controller/controller.go]: My opinion is in-line with @vaikas-google's - having another condition adds more programmatically accessible information on the state of the system.

Regardless, I would like to get beyond this so that @vaikas-google can move forward. Please comment if you are significantly opposed to the following:

- Adding a `AsyncOperationInProgress` condition
- Setting that condition on the instance after the broker returns an async response
- Setting `InstanceConditionReady` after the async operation completes

Comment (pmorie) [inline pkg/controller/controller.go]: > My opinion is in-line with @vaikas-google's - having another condition adds more programmatically accessible information on the state of the system.

I disagree that it adds any additional information.  Here's a ready condition that has all the same information:

```go
InstanceCondition{
  Type: InstanceConditionReady
  Status: ConditionFalse
  Reason: "AsyncProvisionInProgress"
  Message:  "The instance is currently being provisioned"
}
```

Do you disagree that this is equivalent?

Comment (vaikas-google) [inline pkg/controller/controller.go]: and then we need AsyncUpdateInProgress, AsyncDeprovisionInProgress and since the bindings also are probably going to have that we'll basically be doing ~5 string compares. As I said, I think that's janky but whatever.

Comment (jwforres) [inline pkg/controller/controller.go]: I agree with @pmorie here.  Reason is a machine-readable string.  The "reason" the instance is not ready is that an async provision is in progress. From a client side I want to look for one condition, "is the instance ready".  If that condition is False then I want to check why, Reason is the why.

Comment (vaikas-google) [inline pkg/controller/controller.go]: @jwforres and @pmorie and I just a chat. There was a misunderstanding that Operation is always there, which is not the case according to OSB spec as well as the CF documents. Hence we can NOT use the operation as state for knowing that an async operation is ongoing against a ServiceInstance.
So, we need to store state about this. We can store it in the condition like I had done, but it was decided that instead of storing that state in a Condition, we'll add another field in the InstanceStatus indicating that an async operation is in progress. I'm fine with this approach and I'll change the code accordingly.
@arschles do you have any concerns with this approach?

Comment (arschles) [inline pkg/controller/controller.go]: @vaikas-google @jwforres @pmorie I think that is a good compromise

**Comment by arschles** [comment]
Moving to 0.0.5 since this is still in progress

**Comment by pmorie** [comment]
One main thing: do we really need a separate method to handle polling?  Experience writing controllers like this tells me that's it's much easier long-term to keep all of the logic into a single transaction script.

File: pkg/apis/servicecatalog/types.go [hunk 9]
```diff
@@ -208,6 +208,9 @@ type InstanceSpec struct {
 // InstanceStatus represents the current status of an Instance.
 type InstanceStatus struct {
 	Conditions []InstanceCondition
+	// AsyncOpInProgress is set to true if there is an ongoing async operation
+	// against this Service Instance in progress.
+	AsyncOpInProgress bool
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: Do we need to also have a `*string` field for `BrokerProvidedOperationID`?

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go [hunk 10]
```diff
@@ -138,7 +139,7 @@ func (c *openServiceBrokerClient) GetCatalog() (*brokerapi.Catalog, error) {
 	return &catalog, nil
 }
 
-func (c *openServiceBrokerClient) CreateServiceInstance(ID string, req *brokerapi.CreateServiceInstanceRequest) (*brokerapi.CreateServiceInstanceResponse, error) {
+func (c *openServiceBrokerClient) CreateServiceInstance(ID string, req *brokerapi.CreateServiceInstanceRequest) (*brokerapi.CreateServiceInstanceResponse, int, error) {
 	var serviceInstanceURL string
 
 	if req.AcceptsIncomplete {
```
Comment (jmrodri) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: Do you ever expect to work with brokers that don't use async?

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go (hunk 3 above)
Comment (vaikas-google) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: yes

File: pkg/apis/servicecatalog/types.go (hunk 9 above)
Comment (vaikas-google) [inline pkg/apis/servicecatalog/types.go]: Well, it's in the Spec currently (as OSBLastOperation or something like that) but I need to read and see if it should be moved here, I think it's the wrong place for it, but I haven't had time to delve into the details about that move. I'd rather do that in a follow on PR as this is getting large enough already, and I'd rather move this forward.

File: pkg/brokerapi/openservicebroker/open_service_broker_client_test.go [hunk 11]
```diff
@@ -344,6 +402,109 @@ func TestUnbindGone(t *testing.T) {
...
+	_, err := c.PollServiceInstance(testServiceInstanceID, r)
+	if err == nil {
+		t.Fatal("PollServiceInstance did not fail with statusBadRequest")
+	}
+	if !strings.Contains(err.Error(), "400") {
+		t.Fatalf("Did not find the expected error message '400' error: %s", err)
+	}
+}
+
+func TestPollServiceInstanceWithSuccess(t *testing.T) {
```
Comment (vaikas-google) [inline pkg/brokerapi/openservicebroker/open_service_broker_client_test.go]: yeah, as I said there's bunch of tests that I'm finishing up. I'll let y'all know when it's ready for a review.

**Comment by vaikas-google** [comment]
@pmorie adding more code into a single method (notice I already pulled out some stuff out of there) that's already was pushing 400 lines, adding more if/else serial logic seems totally wrong to me. It's very difficult already to test much of this stuff because the methods are so large and breaking into smaller more manageable chunks seems like a much better approach both for understanding the code base, and for being able to write more focused unit tests.Note that here's nothing preventing us from adding a call from  reconcileInstance to pollServiceInstance method later from the reconcileInstance if that's what we want to d since  it's pulled into it's own method. So, yes, imho we really do need to create a separate method that actually handles the polling so that it can be tested more easily instead of adding more logic into one giant serial reconcile method.

**Comment by vaikas-google** [comment]
Ok, I want to add couple of more tests for deprovisioning, but please take another look.

File: pkg/controller/controller.go [hunk 12]
```diff
@@ -1370,6 +1443,144 @@ func (c *controller) bindingDelete(obj interface{}) {
...
+			v1alpha1.BindingConditionReady,
+			v1alpha1.ConditionFalse,
+			errorNonexistentServiceClassReason,
+			"The binding references a ServiceClass that does not exist. "+s,
+		)
+		c.recorder.Event(binding, api.EventTypeWarning, "ReferencesNonexistentServiceClass", s)
+		return nil, nil, "", nil, err
+	}
+
+	servicePlan := findServicePlan(instance.Spec.PlanName, serviceClass.Plans)
```
Comment (pmorie) [inline pkg/controller/controller.go]: isn't the rest of this method just `getServiceClassPlanAndBroker` ?

**Comment by pmorie** [comment]
@vaikas-google LGTM, but tests failed

File: pkg/controller/controller.go (hunk 12 above)
Comment (vaikas-google) [inline pkg/controller/controller.go]: No, because it updates the bindings condition instead of the instance. There's probably a way to combine the two in a clever manner, but for now they are separate.

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go (hunk 10 above)
Comment (vaikas-google) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: Actually, I was going to create another PR later on to remove it, I stumbled on this and was thinking of removing it, but did not want to change that behaviour yet. I wanted to spend little more time making sure I didn't miss anything in the spec on the significance here, but yes, I think we should remove it.

File: pkg/controller/controller.go (hunk 12 above)
Comment (pmorie) [inline pkg/controller/controller.go]: Ah, derp.

**Comment by MHBauer** [comment]
@vaikas-google are you comfortable folding #732 & #733 into this? I don't want to double work on those if you intend to do it. I also don't want to make a bunch of merge collisions later. What do you think?

**Comment by pmorie** [comment]
@MHBauer 

> @vaikas-google are you comfortable folding #732 & #733 into this? I don't want to double work on those if you intend to do it. I also don't want to make a bunch of merge collisions later. What do you think?

Ville and I talked a lot about this yesterday, and I think this PR is really, really big, and pushes things significantly in the right direction.  I'd personally prefer to get this in as-is and do some immediate follow ups (of which there are others we want to do besides the ones you named).

**Comment by pmorie** [comment]
Also, I have people effectively blocked by this that need it.

**Comment by MHBauer** [comment]
Sounds good to me. That's where I'm about.

**Comment by vaikas-google** [comment]
I'm not sure what you mean fold them in. I think last_operation is now handled by this PR correctly. Also the polling interval is now changeable and uses exp. backoff and I think we could make it configurable, but would rather not fold into this pr.

**Comment by pmorie** [comment]
@MHBauer does that mean your are LGTM on this PR?

**Comment by pmorie** [comment]
I have a couple datapoints from @jmrodri and @spadgett that this in conjunction with #757 works correctly for async provisioning, so I think we should merge this.

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 13]
```diff
@@ -209,6 +209,10 @@ type InstanceSpec struct {
 // InstanceStatus represents the current status of an Instance.
 type InstanceStatus struct {
 	Conditions []InstanceCondition `json:"conditions"`
+
+	// AsyncOpInProgress is set to true if there is an ongoing async operation
+	// against this Service Instance in progress.
+	AsyncOpInProgress bool `json:"asyncOpInProgress"`
```
Comment (kibbles-n-bytes) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Just wondering why this is a separate entity from an `InstanceCondition`?

File: pkg/brokerapi/broker_client.go [hunk 14]
```diff
@@ -44,16 +44,18 @@ type InstanceClient interface {
 	// TODO: these should return appropriate response objects (https://github.com/kubernetes-incubator/service-catalog/issues/116).
 
 	// CreateServiceInstance creates a service instance in the respective broker.
-	// This method handles all asynchronous request handling.
-	CreateServiceInstance(ID string, req *CreateServiceInstanceRequest) (*CreateServiceInstanceResponse, error)
+	CreateServiceInstance(ID string, req *CreateServiceInstanceRequest) (*CreateServiceInstanceResponse, int, error)
```
Comment (kibbles-n-bytes) [inline pkg/brokerapi/broker_client.go]: Can we name the `int` return type "responseCode" to improve readability?

File: pkg/brokerapi/openservicebroker/open_service_broker_client.go [hunk 15]
```diff
@@ -40,9 +40,10 @@ const (
...
 	serviceInstanceDeleteFormatString      = "%s/v2/service_instances/%s?service_id=%s&plan_id=%s"
 	serviceInstanceDeleteAsyncFormatString = "%s/v2/service_instances/%s?service_id=%s&plan_id=%s&accepts_incomplete=true"
-	pollingFormatString                    = "%s/v2/service_instances/%s/last_operation"
+	pollingFormatString                    = "%s/v2/service_instances/%s/last_operation?%s"
 	bindingFormatString                    = "%s/v2/service_instances/%s/service_bindings/%s"
 	bindingDeleteFormatString              = "%s/v2/service_instances/%s/service_bindings/%s?service_id=%s&plan_id=%s"
+	queryParamFormatString                 = "%s=%s"
 
 	httpTimeoutSeconds     = 15
 	pollingIntervalSeconds = 1
```
Comment (kibbles-n-bytes) [inline pkg/brokerapi/openservicebroker/open_service_broker_client.go]: `pollingIntervalSeconds` can be removed from here now.

File: pkg/controller/controller.go [hunk 16]
```diff
@@ -262,6 +280,10 @@ const (
 	successBrokerDeletedReason       string = "DeletedSuccessfully"
 	successBrokerDeletedMessage      string = "The broker %v was deleted successfully."
 	successUnboundReason             string = "UnboundSuccessfully"
+	asyncProvisioningReason          string = "Provisioning"
+	asyncProvisioningMessage         string = "The instance is being provisioned asynchronously"
+	asyncDeprovisioningReason        string = "Derovisioning"
```
Comment (kibbles-n-bytes) [inline pkg/controller/controller.go]: Typo with `Derovisioning`

File: pkg/controller/controller.go [hunk 17]
```diff
@@ -1370,6 +1479,144 @@ func (c *controller) bindingDelete(obj interface{}) {
 	glog.V(4).Infof("Received delete event for Binding %v/%v", binding.Namespace, binding.Name)
 }
 
+// getServiceClassPlanAndBroker is a sequence of operations that's done in couple of
```
Comment (kibbles-n-bytes) [inline pkg/controller/controller.go]: This comment seems unnecessarily wordy; I feel like the common use case for a function in general is for "a sequence of operations that's done in a couple of places". I'd just put something like:

```
getServiceClassPlanAndBroker returns the class, plan, broker name, and a newly-constructed broker client for a given instance.
```

File: pkg/controller/controller.go [hunk 18]
```diff
@@ -1370,6 +1479,144 @@ func (c *controller) bindingDelete(obj interface{}) {
...
+// places so this method fetches the Service Class, Service Plan and creates
+// a brokerClient to use for that method given an Instance.
+func (c *controller) getServiceClassPlanAndBroker(instance *v1alpha1.Instance) (*v1alpha1.ServiceClass, *v1alpha1.ServicePlan, string, brokerapi.BrokerClient, error) {
```
Comment (kibbles-n-bytes) [inline pkg/controller/controller.go]: Can we name the `string` return type "brokerName" to improve readability?

File: pkg/controller/controller.go [hunk 19]
```diff
@@ -1370,6 +1479,144 @@ func (c *controller) bindingDelete(obj interface{}) {
...
+
+	glog.V(4).Infof("Creating client for Broker %v, URL: %v", broker.Name, broker.Spec.URL)
+	brokerClient := c.brokerClientCreateFunc(broker.Name, broker.Spec.URL, username, password)
+	return serviceClass, servicePlan, broker.Name, brokerClient, nil
+}
+
+// getServiceClassPlanAndBrokerForBinding is a sequence of operations that's
+// done to validate service plan, service class exist, and handles creating
+// a brokerclient to use for a given Instance.
+func (c *controller) getServiceClassPlanAndBrokerForBinding(instance *v1alpha1.Instance, binding *v1alpha1.Binding) (*v1alpha1.ServiceClass, *v1alpha1.ServicePlan, string, brokerapi.BrokerClient, error) {
```
Comment (kibbles-n-bytes) [inline pkg/controller/controller.go]: Same as above: can we name the `string` return type "brokerName"?

**Comment by kibbles-n-bytes** [comment]
Looks good to me, some small nits but nothing to do with the underlying architecture.

**Comment by MHBauer** [comment]
Don't let me hold it up. I was in the middle of a review but had to go present.

I've got a broker I can try with this.

**Comment by pmorie** [comment]
I'm going to merge this in good faith because we have good data from @jmrodri and @spadgett that this and #757 collectively make async much better.
